README
SmartOptimizer v1.8
------------------------
Author: Ali Farhadi (http://farhadi.ir/)
Released under the terms of the GNU Public License.
See the GPL for details (http://www.gnu.org/licenses/gpl.html).


What is SmartOptimizer?
-----------------------
SmartOptimizer is a php program that enhances your website performance using
techniques such as minifying, compression, caching, concatenation and embedding.
The idea is based on yahoo performace rules (http://developer.yahoo.com/performance/rules.html).


SmartOptimizer Features
-----------------------
* SmartOptimizer Increases your website download speed by reducing the size of JavaScript and
  CSS files approximately up to 80% or even higher.
* Reduces count of http requests by combining js/css files together.
* Also reduces count of http requests by embedding css images inside css files.
* SmartOptimizer generates minified, compressed, combined and embedded files on the
  first request and caches them on the server and uses the cache on future requests.
* It doesn't embed duplicated urls (i.e. css sprites) in css files for better performance.
* It also knows to regenerate already cached files if the originals have been changed.
* SmartOptimizer handles browsers that doesn't support gzip encoding and sends them non-gzipped content.
* It also forces the browser to cache the files optionally forever or until they have not changed.
* SmartOptimizer removes Etag headers. (for better performance on clustered servers).
* It is easy to install, since it requires no code modifications whatsoever. (other than .htaccess)
* All the features is configurable using config file.


Installation Requirements
-------------------------
* PHP 4.3.0 or higher.
* Apache with mod_rewrite enabled.


Installation Instructions
------------------------
1. Upload smartoptimizer folder to your website.
2. "smartoptimizer/cache" folder should be writable (in most cases 777 permission is needed).
3. Upload .htaccess file on the folder that smartoptimizer folder is placed, if there is no another htaccess file currently.
   Otherwise, Copy content of SmartOptimizer htaccess file and paste it inside your current htaccess at the beginning of the file.
4. It's ready. Enjoy it.


Bug reports
-----------
SmartOptimizer is in beta stage and needs a large scale testing specially for its css and js minifiers.
So any feedback is greatly appreciated.

Use the issue tracker at github for bug reports:
http://github.com/farhadi/SmartOptimizer/issues

CHANGELOG
NOTE: I'm not going to update this file any more. Use git-log instead.

-------------------- 1.7 beta [2008-12-06] ---------------------
2008-12-16  Ali Farhadi <ali@farhadi.ir>
	* ability to ignore some file types (e.g. htc) from being embedded through embedExceptions config.
2008-12-06  Ali Farhadi <ali@farhadi.ir>
	* minifiers/css.php: Issue #2 fixed (problem with mime_content_type function).
	* config.php: gzip compression level can be set in config file.
	* index.php: Content-Type of js files was changed for compatibility with all of browsers. (from application/x-javascript to text/javascript).

-------------------- 1.6 beta [2008-06-13] ---------------------
2008-06-13  Ali Farhadi <ali@farhadi.ir>
	Change log and brief documentation added.
	* minifiers/css.php, index.php: Ability of embedding css images and css imported files was added.
	* minifiers/css.php: Duplicated urls (i.e. css sprites) will not be embeded for better performance.
	* config.php: embed feature can be disabled in config file.
	* config.php: Maximum file size of embedded files can be specified.
	* .htaccess: bug fixed in removing Etag headers for static contents.

-------------------- 1.5 beta [2008-05-24] ---------------------
2008-05-24  Ali Farhadi <ali@farhadi.ir>
	The project renamed to SmartOptimizer.
	SmartOptimizer is a completely rewritten and restructured version of JSmart.
	* index.php: Ability of concatenating js and css files was added.
	* minifiers/css.php: Relative paths in css files will be corrected if needed.
	* config.php: Cache settings is more customizable.
	* config.php: Ability of disabling minification and gzip compression was added.
	* index.php: Handling expire headers for a wide range of file types was added.
	* index.php: getallheaders function is not used anymore, since it is not supported in all the servers.

-------------------- 1.0 Final Release [2006-07-07] ---------------------
2006-07-07  Ali Farhadi <ali@farhadi.ir>
	* load.php: Ability of compressing CSS files was added.
	* load.php: Content-Type of js files was corrected. (from text/plain to application/x-javascript)
    
-------------------- 1.0 RC2 [2006-07-01] ---------------------
2006-07-01  Ali Farhadi <ali@farhadi.ir>
	* config.php: Charset of js files can be specified in config file.
	* load.php: Debugging capabilities was added.
	* config.php: Debug mode can be enabled or disabled using config file.
	* config.php: Cache feature can be disabled in config file
	
-------------------- 1.0 RC [2006-06-29] ---------------------
2006-06-29  Ali Farhadi <ali@farhadi.ir>
	* config.php: JSmart is more customizable using config.php
	* load.php: JSmart will not crash if zlib extension is not installed.
	
-------------------- 1.0 beta [2006-06-25] ---------------------
2006-06-25  Ali Farhadi <ali@farhadi.ir>
	It is the first release


smartoptimizer/config.php
<?php
/*
 * SmartOptimizer Configuration File
 */

//base dir (a relative path to the base directory)
$settings['baseDir'] = '';

//Encoding of your js and css files. (utf-8 or iso-8859-1)
$settings['charSet'] = 'utf-8'; 

//Show error messages if any error occurs (true or false)
$settings['debug'] = true;

//use this to set gzip compression On or Off
$settings['gzip'] = true;

//use this to set gzip compression level (an integer between 1 and 9)
$settings['compressionLevel'] = 9;

//these types of files will not be gzipped nor minified
$settings['gzipExceptions'] = array('gif','jpeg','jpg','png','swf'); 

//use this to set Minifier On or Off
$settings['minify'] = true;

//use this to set file concatenation On or Off
$settings['concatenate'] = true;

//separator for files to be concatenated. Tip: Don't use '.' if you use groups and name them like suggested.
$settings['separator'] = ',';

//specifies whether to emebed files included in css files using the data URI scheme or not 
$settings['embed'] = true;

//The maximum size of an embedded file. (use 0 for unlimited size)
$settings['embedMaxSize'] = 5120; //5KB

//these types of files will not be embedded
$settings['embedExceptions'] = array('htc'); 

//to set server-side cache On or Off
$settings['serverCache'] = true;

//if you change it to false, the files will not be checked for modifications and always cached files will be used (for better performance)
$settings['serverCacheCheck'] = true;

//cache dir
$settings['cacheDir'] = 'cache/';

//prefix for cache files
$settings['cachePrefix'] = 'so_';

//to set client-side cache On or Off
$settings['clientCache'] = true;

//Setting this to false will force the browser to use cached files without checking for changes.
$settings['clientCacheCheck'] = false;

//Minifier to use when parsing js files. Add yours in /minifiers directory, and implement function minify_js($text_to_minify)
$settings['jsMinifier'] = 'packer';

//Minifier to use when parsing css files. Add yours in /minifiers directory, and implement function minify_css($text_to_minify)
$settings['cssMinifier'] = 'css';

//To use groups effectively, include your files with absolute paths. Define constants for easy access. 
define('MY_JS_DIR', $_SERVER['DOCUMENT_ROOT'] . '/tkitt/js/');
define('MY_CSS_DIR', $_SERVER['DOCUMENT_ROOT'] . '/tkitt/css/');

//groups configuration. Call a group by using /path/to/smartoptimizer/?group.group_name.
//Tip: If you use 'test.js' as your group name, you'll get a nifty file name like 'group.test.js' 
$settings['groups'] = array(
	'test.js' => array(MY_JS_DIR.'test.js', MY_JS_DIR.'test2.js'),
	'test.css' => array(MY_JS_DIR.'style.css', MY_JS_DIR.'print.css')
);

smartoptimizer/minifiers/js.php
<?php
/*
 * SmartOptimizer JavaScript Minifier
 */
function minify_js($str) {
	$res = '';
	$maybe_regex = true;
	$i=0;
	$current_char = '';
	while ($i+1<strlen($str)) {
		if ($maybe_regex && $str[$i]=='/' && $str[$i+1]!='/' && $str[$i+1]!='*' && @$str[$i-1]!='*') {//regex detected
			if (strlen($res) && $res[strlen($res)-1] === '/') $res .= ' ';
			do {
				if ($str[$i] == '\\') {
					$res .= $str[$i++];
				} elseif ($str[$i] == '[') {
					do {
						if ($str[$i] == '\\') {
							$res .= $str[$i++];
						}
						$res .= $str[$i++];
					} while ($i<strlen($str) && $str[$i]!=']');
				}
				$res .= $str[$i++];
			} while ($i<strlen($str) && $str[$i]!='/');
			$res .= $str[$i++];
			$maybe_regex = false;
			continue;
		} elseif ($str[$i]=='"' || $str[$i]=="'") {//quoted string detected
			$quote = $str[$i];
			do {
				if ($str[$i] == '\\') {
					$res .= $str[$i++];
				}
				$res .= $str[$i++];
			} while ($i<strlen($str) && $str[$i]!=$quote);
			$res .= $str[$i++];
			continue;
		} elseif ($str[$i].$str[$i+1]=='/*' && @$str[$i+2]!='@') {//multi-line comment detected
			$i+=3;
			while ($i<strlen($str) && $str[$i-1].$str[$i]!='*/') $i++;
			if ($current_char == "\n") $str[$i] = "\n";
			else $str[$i] = ' ';
		} elseif ($str[$i].$str[$i+1]=='//') {//single-line comment detected
			$i+=2;
			while ($i<strlen($str) && $str[$i]!="\n" && $str[$i]!="\r") $i++;
		}
		


		$LF_needed = false;
		if (preg_match('/[\n\r\t ]/', $str[$i])) {
			if (strlen($res) && preg_match('/[\n ]/', $res[strlen($res)-1])) {
				if ($res[strlen($res)-1] == "\n") $LF_needed = true;
				$res = substr($res, 0, -1);
			}
			while ($i+1<strlen($str) && preg_match('/[\n\r\t ]/', $str[$i+1])) {
				if (!$LF_needed && preg_match('/[\n\r]/', $str[$i])) $LF_needed = true;
				$i++;
			}
		}
		
		if (strlen($str) <= $i+1) break;
		
		$current_char = $str[$i];
		
		if ($LF_needed) $current_char = "\n";
		elseif ($current_char == "\t") $current_char = " ";
		elseif ($current_char == "\r") $current_char = "\n";
		
		// detect unnecessary white spaces
		if ($current_char == " ") {
			if (strlen($res) &&
				(
				preg_match('/^[^(){}[\]=+\-*\/%&|!><?:~^,;"\']{2}$/', $res[strlen($res)-1].$str[$i+1]) ||
				preg_match('/^(\+\+)|(--)$/', $res[strlen($res)-1].$str[$i+1]) // for example i+ ++j;
				)) $res .= $current_char;
		} elseif ($current_char == "\n") {
			if (strlen($res) &&
				(
				preg_match('/^[^({[=+\-*%&|!><?:~^,;\/][^)}\]=+\-*%&|><?:,;\/]$/', $res[strlen($res)-1].$str[$i+1]) ||
				(strlen($res)>1 && preg_match('/^(\+\+)|(--)$/', $res[strlen($res)-2].$res[strlen($res)-1])) ||
				(strlen($str)>$i+2 && preg_match('/^(\+\+)|(--)$/', $str[$i+1].$str[$i+2])) ||
				preg_match('/^(\+\+)|(--)$/', $res[strlen($res)-1].$str[$i+1])// || // for example i+ ++j;
				)) $res .= $current_char;
		} else $res .= $current_char;
		
		// if the next charachter be a slash, detects if it is a divide operator or start of a regex
		if (preg_match('/[({[=+\-*\/%&|!><?:~^,;]/', $current_char)) $maybe_regex = true;
		elseif (!preg_match('/[\n ]/', $current_char)) $maybe_regex = false;
		
		$i++;
	}
	if ($i<strlen($str) && preg_match('/[^\n\r\t ]/', $str[$i])) $res .= $str[$i];
	return $res;
}
?>

smartoptimizer/minifiers/css.php
<?php
/*
 * SmartOptimizer CSS Minifier
 */

function convertUrl($url, $count)
{
	global $settings, $mimeTypes, $fileDir;
	
	static $baseUrl = '';
	
	$url = trim($url);
	
	if (preg_match('@^[^/]+:@', $url)) return $url;
	
	$fileType = substr(strrchr($url, '.'), 1);
	if (isset($mimeTypes[$fileType])) $mimeType = $mimeTypes[$fileType];
	elseif (function_exists('mime_content_type')) $mimeType = mime_content_type($url);
	else $mimeType = null;
	
	if (!$settings['embed'] ||
		!file_exists($fileDir.$url) ||
		($settings['embedMaxSize'] > 0 && filesize($fileDir.$url) > $settings['embedMaxSize']) ||
		!$fileType ||
		!$mimeType ||
		$count > 1) {
		if (strpos($_SERVER['REQUEST_URI'], $_SERVER['SCRIPT_NAME'].'?') === 0 ||
			strpos($_SERVER['REQUEST_URI'], rtrim(dirname($_SERVER['SCRIPT_NAME']), '\/').'/?') === 0) {
			if (!$baseUrl) return $fileDir . $url;
		}
		return $baseUrl . $url;
	}
	
	$contents = file_get_contents($fileDir.$url);
	 
	if ($fileType == 'css') {
		$oldFileDir = $fileDir;
		$fileDir = rtrim(dirname($fileDir.$url), '\/').'/';
		$oldBaseUrl = $baseUrl;
		$baseUrl = 'http'.(@$_SERVER['HTTPS']?'s':'').'://'.$_SERVER['HTTP_HOST'].rtrim(dirname($_SERVER['SCRIPT_NAME']), '\/').'/'.$fileDir;
		$contents = minify_css($contents);		
		$fileDir = $oldFileDir;
		$baseUrl = $oldBaseUrl;
	}
	
	$base64   = base64_encode($contents); 
	return 'data:' . $mimeType . ';base64,' . $base64;
}

function minify_css($str) {
	$res = '';
	$i=0;
	$inside_block = false;
	$current_char = '';
	while ($i+1<strlen($str)) {
		if ($str[$i]=='"' || $str[$i]=="'") {//quoted string detected
			$res .= $quote = $str[$i++];
			$url = '';
			while ($i<strlen($str) && $str[$i]!=$quote) {
				if ($str[$i] == '\\') {
					$url .= $str[$i++];
				}
				$url .= $str[$i++];
			}
			if (strtolower(substr($res, -5, 4))=='url(' || strtolower(substr($res, -9, 8)) == '@import ') {
				$url = convertUrl($url, substr_count($str, $url));
			}
			$res .= $url;
			$res .= $str[$i++];
			continue;
		} elseif (strtolower(substr($res, -4))=='url(') {//url detected
			$url = '';
			do {
				if ($str[$i] == '\\') {
					$url .= $str[$i++];
				}
				$url .= $str[$i++];
			} while ($i<strlen($str) && $str[$i]!=')');
			$url = convertUrl($url, substr_count($str, $url));
			$res .= $url;
			$res .= $str[$i++];
			continue;
		} elseif ($str[$i].$str[$i+1]=='/*') {//css comment detected
			$i+=3;
			while ($i<strlen($str) && $str[$i-1].$str[$i]!='*/') $i++;
			if ($current_char == "\n") $str[$i] = "\n";
			else $str[$i] = ' ';
		}
		
		if (strlen($str) <= $i+1) break;
		
		$current_char = $str[$i];
		
		if ($inside_block && $current_char == '}') {
			$inside_block = false;
		}
		
		if ($current_char == '{') {
			$inside_block = true;
		}
		
		if (preg_match('/[\n\r\t ]/', $current_char)) $current_char = " ";
		
		if ($current_char == " ") {
			$pattern = $inside_block?'/^[^{};,:\n\r\t ]{2}$/':'/^[^{};,>+\n\r\t ]{2}$/';
			if (strlen($res) &&	preg_match($pattern, $res[strlen($res)-1].$str[$i+1]))
				$res .= $current_char;
		} else $res .= $current_char;
		
		$i++;
	}
	if ($i<strlen($str) && preg_match('/[^\n\r\t ]/', $str[$i])) $res .= $str[$i];
	return $res;
}
?>

smartoptimizer/index.php
<?php
/* SmartOptimizer v1.8
 * SmartOptimizer enhances your website performance using techniques
 * such as compression, concatenation, minifying, caching, and embedding on demand.
 *
 * Copyright (c) 2006-2010 Ali Farhadi (http://farhadi.ir/)
 * Released under the terms of the GNU Public License.
 * See the GPL for details (http://www.gnu.org/licenses/gpl.html).
 *
 * Author: Ali Farhadi (a.farhadi@gmail.com)
 * Website: http://farhadi.ir/
 */

//Default settings
$settings = array(
	'baseDir' => '../',
	'charSet' => 'utf-8',
	'debug' => true,
	'gzip' => true,
	'compressionLevel' => 9,
	'gzipExceptions' => array('gif','jpeg','jpg','png','swf','ico'),
	'minify' => true,
	'concatenate' => true,
	'separator' => ',',
	'embed' => true,
	'embedMaxSize' => 5120,
	'embedExceptions' => array('htc'),
	'serverCache' => true,
	'serverCacheCheck' => false,
	'cacheDir' => 'cache/',
	'cachePrefix' => 'so_',
	'clientCache' => true,
	'clientCacheCheck' => false
);

//mime types
$mimeTypes = array(
	"js"	=> "text/javascript",
	"css"	=> "text/css",
	"htm"	=> "text/html",
	"html"	=> "text/html",
	"xml"	=> "text/xml",
	"txt"	=> "text/plain",
	"jpg"	=> "image/jpeg",
	"jpeg"	=> "image/jpeg",
	"png"	=> "image/png",
	"gif"	=> "image/gif",
	"swf"	=> "application/x-shockwave-flash",
	"ico"	=> "image/x-icon",
);

function headerExit($status) {
	header("HTTP/1.0 $status");
	exit();
}

function headerNoCache() {
	// already expired
	header("Expires: Mon, 26 Jul 1997 05:00:00 GMT");

	// always modified
	header("Last-Modified: " . gmdatestr());

	// HTTP/1.1
	header("Cache-Control: no-store, no-cache, must-revalidate");
	header("Cache-Control: post-check=0, pre-check=0", false);
	header("Cache-Control: max-age=0", false);

	// HTTP/1.0
	header("Pragma: no-cache");

	//generate a unique Etag each time
	header('Etag: '.microtime());
}

function headerNeverExpire(){
	header("Expires: " . gmdatestr(time() + 315360000));
	header("Cache-Control: max-age=315360000");
}

function debugExit($msg){
	global $settings;
	if (!$settings['debug']) {
		headerExit('404 Not Found');
	}
	headerNoCache();
	header('Content-Type: text/html; charset='.$settings['charSet']);
	header("Content-Encoding: none");
	echo "//<script>\n";
	echo "alert('SmartOptimizer Error: ".str_replace("\n", "\\n", addslashes($msg))."');\n";
	echo "//</script>\n";
	exit();
}

function gmdatestr($time = null) {
	if (is_null($time)) $time = time();
	return gmdate("D, d M Y H:i:s", $time) . " GMT";
}

function filesmtime() {
	global $files, $fileType;
	static $filesmtime;
	if ($filesmtime) return $filesmtime;
	$filesmtime = max(@filemtime("minifiers/$fileType.php"), filemtime('index.php'), filemtime('config.php'));
	foreach ($files as $file) {
		if (!file_exists($file)) debugExit("File not found ($file).");
		$filesmtime = max(filemtime($file), $filesmtime);
	}
	return $filesmtime;
}

@include('config.php');

list($query) = explode('?', urldecode($_SERVER['QUERY_STRING']));

if(isset($settings['groups']) && substr($query, 0, 6) == 'group.'){
	//we have a pre-defined group to include
	$group_name = str_replace('group.','',$query);
	if(isset($settings['groups'][$group_name])) {
		$fileNames = $settings['groups'][$group_name];
		$fileDir = '';
	}
	else debugExit("Group ($group_name) not set. Please edit config.");
} else {
	if (preg_match('/^\/?(.+\/)?(.+)$/', $query, $matchResult)) {
		$fileNames = $matchResult[2];
		$fileDir = $settings['baseDir'].$matchResult[1];
	} else debugExit("Invalid file name ($query)");
} 

//if (strpos(realpath($fileDir), realpath($settings['baseDir'])) !== 0) debugExit("File is out of base directory.");

if ($settings['concatenate']) {
	if(!is_array($fileNames)){
		$files = explode($settings['separator'], $fileNames);
	}
	else{
		$files = $fileNames;
	}
	$settings['concatenate'] = count($files) > 1;
} else $files = array($fileNames);

foreach ($files as $key => $file) {
	if (preg_match('/^[^\x00]+\.([a-z0-9]+)$/i', $file, $matchResult)) {
		$fileTypes[] = strtolower($matchResult[1]);
	} else debugExit("Unsupported file ($file)");

	$files[$key] = $fileDir.$file;
}

if ($settings['concatenate']) {
	if (count(array_unique($fileTypes)) > 1) debugExit("Files must be of the same type.");
}

$fileType = $fileTypes[0];

if (!isset($mimeTypes[$fileType])) debugExit("Unsupported file type ($fileType)");
header("Content-Type: {$mimeTypes[$fileType]}; charset=".$settings['charSet']);

$settings['gzip'] =
	($settings['gzip'] &&
	!in_array($fileType, $settings['gzipExceptions']) &&
	in_array('gzip', array_map('trim', explode(',' , @$_SERVER['HTTP_ACCEPT_ENCODING']))) &&
	function_exists('gzencode'));

if ($settings['gzip']) header("Content-Encoding: gzip");

$settings['minify'] = $settings['minify'] && file_exists('minifiers/'.$fileType.'.php');
$settings['embed'] = $settings['embed'] && $fileType == 'css' && (!preg_match('/msie/i', $_SERVER['HTTP_USER_AGENT']) || preg_match('/msie 8|opera/i', $_SERVER['HTTP_USER_AGENT']));
$settings['serverCache'] = $settings['serverCache'] && ($settings['minify'] || $settings['gzip'] || $settings['concatenate'] || $settings['embed']);

if ($settings['serverCache']) {
	$cachedFile = $settings['cacheDir'].$settings['cachePrefix'].md5($query.($settings['embed']?'1':'0')).'.'.$fileType.($settings['gzip'] ? '.gz' : '');
}

$generateContent = ((!$settings['serverCache'] && (!$settings['clientCache'] || !$settings['clientCacheCheck'] || !isset($_SERVER['HTTP_IF_MODIFIED_SINCE']) || $_SERVER['HTTP_IF_MODIFIED_SINCE'] != gmdatestr(filesmtime()))) ||
	($settings['serverCache'] && (!file_exists($cachedFile) || ($settings['serverCacheCheck'] && filesmtime() > filemtime($cachedFile)))));

if ($settings['clientCache'] && $settings['clientCacheCheck']) {
	if ($settings['serverCache'] && !$generateContent) $mtime = filemtime($cachedFile);
	elseif ($settings['serverCache']) $mtime = time();
	else $mtime = filesmtime();
	$mtimestr = gmdatestr($mtime);
}

if (!$settings['clientCache'] || !$settings['clientCacheCheck'] || !isset($_SERVER['HTTP_IF_MODIFIED_SINCE']) || $_SERVER['HTTP_IF_MODIFIED_SINCE'] != $mtimestr) {
	if ($settings['clientCache'] && $settings['clientCacheCheck']) {
		header("Last-Modified: " . $mtimestr);
		header("Cache-Control: must-revalidate");
	} elseif ($settings['clientCache']) {
		headerNeverExpire();
	} else headerNoCache();

	if ($generateContent) {
		if ($settings['minify']) include('minifiers/'.$fileType.'.php');
		$content = array();
		foreach ($files as $file) (($content[] = @file_get_contents($file)) !== false) || debugExit("File not found ($file).");
		$content = implode("\n", $content);
		if ($settings['minify']) $content = call_user_func('minify_' . $fileType, $content);
		if ($settings['gzip']) $content = gzencode($content, $settings['compressionLevel']);
		if ($settings['serverCache']) {
			$handle = @fopen($cachedFile, 'w') or debugExit("Could not create cache file($cachedFile).");
			fwrite($handle, $content);
			fclose($handle);
		}
		header('Content-Length: ' . strlen($content));
		echo $content;
	} else {
		header('Content-Length: ' . filesize($cachedFile));
		readfile($cachedFile);
	}
} else headerExit('304 Not Modified');

?>